contracts/tokens/ERC721/N-ERC721.cairo

```diff
@@ -0,0 +1,293 @@
+
+@external
+func approve{
+    syscall_ptr: felt*,
+    pedersen_ptr: HashBuiltin*,
+    bitwise_ptr : BitwiseBuiltin*,
+    range_check_ptr
+}(
+    spender: felt,
+    tokenId: felt
+):
+    let (caller) = get_caller_address()
+
+    let (owner) = _owners.read(tokenId)
+    tempvar callerIsOwner = 0 #false by default
+    if caller == owner:
+        callerIsOwner = 1
+    end
+    let (approved) = _isApprovedForAll.read(owner, caller)
+    let (can_approve) = bitwise_or(callerIsOwner, approved)
+    assert can_approve = 1
+
+    _tokenApprovals.write(tokenId, spender)
+    return ()
+end
+
+@external
+func setApprovalForAll{
+    syscall_ptr: felt*,
+    pedersen_ptr: HashBuiltin*,
+    range_check_ptr
+}(
+    operator: felt,
+    approved: felt
+):
+    let (caller) = get_caller_address()
+    _isApprovedForAll.write(caller, operator, approved)
+    return ()
+end
+
+@external
+func transferFrom{
+    syscall_ptr: felt*,
+    pedersen_ptr: HashBuiltin*,
+    bitwise_ptr : BitwiseBuiltin*,
+    range_check_ptr
+}(
+    sender: felt,
+    recipient: felt,
+    tokenId: felt
+):
+    let (caller) = get_caller_address()
+    let (owner) = _owners.read(tokenId)
+
+    assert sender = owner #wrong sender
+
+    assert_not_zero(recipient)
+
+    tempvar isCallerOwner = 0
+    if owner == caller:
+        isCallerOwner = 1
+    end
+    let (approvedSpender) = _tokenApprovals.read(tokenId)
+    tempvar isApproved = 0
+    if approvedSpender == caller:
+        isApproved = 1
+    end
+    let (isApprovedForAll) = _isApprovedForAll.read(owner, caller)
+    let (canTransfer1) = bitwise_or(isCallerOwner, isApproved)
+    let (canTransfer) = bitwise_or(canTransfer1, isApprovedForAll)
+    assert canTransfer = 1
+
+    let (previousOwnerBalance) = _balances.read(sender)
+    let (newOwnerBalance) = _balances.read(recipient)
+
+    _balances.write(sender, previousOwnerBalance - 1)
+    _balances.write(recipient, newOwnerBalance + 1)
+
+    _owners.write(tokenId, recipient)
+
+    _tokenApprovals.write(tokenId, 0)
+
+    return ()
+end
+
+#############################################
+##             INTERNAL LOGIC              ##
+#############################################
+
+func _mint{
+    syscall_ptr: felt*,
+    pedersen_ptr: HashBuiltin*,
+    range_check_ptr
+}(
+    recipient: felt,
+    tokenId: felt
+):
+    assert_not_zero(recipient) #invalid recipient
+    let (token_owner) = _owners.read(tokenId)
+    assert token_owner = 0 #already minted
+
+    let (currentBalance) = _balances.read(owner=recipient)
+    _balances.write(recipient, currentBalance + 1)
+
+    let (currentSupply) = _totalSupply.read()
+    _totalSupply.write(currentSupply + 1)
+
+    _owners.write(tokenId, recipient)
+
+    return ()
+end
+
+func _burn{
+    syscall_ptr: felt*,
+    pedersen_ptr: HashBuiltin*,
+    range_check_ptr
+}(
+    tokenId: felt
+):
+    let (owner) = _owners.read(tokenId)
+    assert_not_equal(owner, 0) #not minted
```

Comment:
might just want to use `assert_not_zero`, equality checks seem more expensive in cairo-lang

https://github.com/starkware-libs/cairo-lang/blob/3d33c4e829a87bc3d88cf04ed6a489e788918b8b/src/starkware/cairo/common/math.cairo#L3-L36

![Screenshot from 2022-01-02 22-26-38](https://user-images.githubusercontent.com/94925499/147897790-40397f90-b0a3-445b-a158-afe092bff9ec.png)